modules/services/gpg-agent.nix

```diff
@@ -256,79 +275,93 @@ in {
+          };
 
-        Install = { WantedBy = [ "sockets.target" ]; };
-      };
-    }
+          Socket = {
+            ListenStream = gpgconf "S.gpg-agent";
+            FileDescriptorName = "std";
+            SocketMode = "0600";
+            DirectoryMode = "0700";
+          };
 
-    (mkIf cfg.enableSshSupport {
-      systemd.user.sockets.gpg-agent-ssh = {
-        Unit = {
-          Description = "GnuPG cryptographic agent (ssh-agent emulation)";
-          Documentation =
-            "man:gpg-agent(1) man:ssh-add(1) man:ssh-agent(1) man:ssh(1)";
+          Install = { WantedBy = [ "sockets.target" ]; };
         };
-
-        Socket = {
-          ListenStream = gpgconf "S.gpg-agent.ssh";
-          FileDescriptorName = "ssh";
-          Service = "gpg-agent.service";
-          SocketMode = "0600";
-          DirectoryMode = "0700";
+      }
+
+      (mkIf cfg.enableSshSupport {
+        systemd.user.sockets.gpg-agent-ssh = {
+          Unit = {
+            Description = "GnuPG cryptographic agent (ssh-agent emulation)";
+            Documentation =
+              "man:gpg-agent(1) man:ssh-add(1) man:ssh-agent(1) man:ssh(1)";
+          };
+
+          Socket = {
+            ListenStream = gpgconf "S.gpg-agent.ssh";
+            FileDescriptorName = "ssh";
+            Service = "gpg-agent.service";
+            SocketMode = "0600";
+            DirectoryMode = "0700";
+          };
+
+          Install = { WantedBy = [ "sockets.target" ]; };
         };
-
-        Install = { WantedBy = [ "sockets.target" ]; };
-      };
-    })
-
-    (mkIf cfg.enableExtraSocket {
-      systemd.user.sockets.gpg-agent-extra = {
-        Unit = {
-          Description =
-            "GnuPG cryptographic agent and passphrase cache (restricted)";
-          Documentation = "man:gpg-agent(1) man:ssh(1)";
+      })
+
+      (mkIf cfg.enableExtraSocket {
+        systemd.user.sockets.gpg-agent-extra = {
+          Unit = {
+            Description =
+              "GnuPG cryptographic agent and passphrase cache (restricted)";
+            Documentation = "man:gpg-agent(1) man:ssh(1)";
+          };
+
+          Socket = {
+            ListenStream = gpgconf "S.gpg-agent.extra";
+            FileDescriptorName = "extra";
+            Service = "gpg-agent.service";
+            SocketMode = "0600";
+            DirectoryMode = "0700";
+          };
+
+          Install = { WantedBy = [ "sockets.target" ]; };
         };
-
-        Socket = {
-          ListenStream = gpgconf "S.gpg-agent.extra";
-          FileDescriptorName = "extra";
-          Service = "gpg-agent.service";
-          SocketMode = "0600";
-          DirectoryMode = "0700";
+      })
+    ]))
+
+    (mkIf pkgs.stdenv.hostPlatform.isDarwin {
+      launchd.agents.gpg-agent = {
+        enable = true;
+        config = {
+          ProgramArguments = [ "${gpgPkg}/bin/gpg-connect-agent" "/bye" ];
+          RunAtLoad = cfg.enableSshSupport;
```

Comment:
Doesn't this make automatically starting gpg-agent dependent on having SSH support enabled? What if someone wanted to start automatically start GPG agent, but without SSH support (i.e. just enabled services.gpg-agent.?

Sorry if I'm missing something, but I don't see the launchd equivalent of `${gpgPkg}/bin/gpg-agent --supervised` (from the gpg-agent systemd user service) anywhere for Darwin. Is it not necessary due to the socket stuff?

Reply:
I think you're right @ratorx 👍 In my implementation, I simply have `RunAtLoad` set to `true`.

Regarding launching the agent, it appears (at least on my system, with `gpg-agent` 2.4.0), there no longer exists a `--supervised` option:
```
Options used for startup:
     --daemon                           run in daemon mode (background)
     --server                           run in server mode (foreground)
     --no-detach                        do not detach from the console
 -s, --sh                               sh-style command output
 -c, --csh                              csh-style command output
     --options FILE                     read options from FILE
```


I tried launching with what I thought was likely the closest thing: `gpg-agent --server --no-detach` (also tried without `--no-detach`), but seemingly any shell I launhced (which, with the intended & implemented configuration, will launch the agent if one isn't present) wasn't able to pick up the running gpg-agent process, so started its own one.

This is what I'm using to launch the agent, which works (and seems to be what's suggested in most of the literature I can find):
```nix
          ProgramArguments = [ "${gpgPkg}/bin/gpgconf" "--launch" "gpg-agent" ]
            ++ optionals cfg.verbose [ "--verbose" ];
```

The only caveat with this is that the process is only launched by launchd, not supervised in any manner. So, unloading the service will not terminate the disowned process, for instance.  
It doesn't seem there is an option for telling `gpgconf` to launch the `gpg-agent` component in a foreground manner.

But, it works nicely so far, in my experience :)

Reply:
> Doesn't this make automatically starting gpg-agent dependent on having SSH support enabled?

Yeah, I looked into this a while back and that does seem to be the case. I wasn't able to make much progress then, but it looks like @cmacrae has made a lot of progress testing out a fix. @cmacrae do you have a branch based off mine that I could test out? I thought you had opened a PR already but maybe not.